ArduPlane/mode_autoland.cpp

```diff
@@ -152,14 +168,46 @@ bool ModeAutoLand::_enter()
 void ModeAutoLand::update()
 {
     plane.calc_nav_roll();
+
+    // check if initial climb before turn altitude has been reached
+bool terrain_following_active = false;
+#if AP_TERRAIN_AVAILABLE
+    terrain_following_active = plane.terrain_enabled_in_current_mode();
+#endif
+    const float altitude = plane.relative_ground_altitude(false,terrain_following_active); //above terrain if enabled,otherwise above home
+    uint16_t roll_limit_cd;
+    if (!done_climb) {
+        // Constrain the roll limit during climb before turn. This also leaves
+        // some leeway for fighting wind.
+        roll_limit_cd = MIN(plane.roll_limit_cd, plane.g.level_roll_limit*100);
+        plane.nav_roll_cd = constrain_int16(plane.nav_roll_cd, -roll_limit_cd, roll_limit_cd);
+        //climb at TECS max until alt reached
+        plane.target_altitude.amsl_cm = plane.current_loc.alt + climb_before_turn_alt * 100.0f;
+#if AP_TERRAIN_AVAILABLE
+        plane.target_altitude.terrain_alt_cm = plane.target_altitude.amsl_cm;
```

Comment:
I suspect this is key to how why works. It sets the TECS target to a huge altitude which is enough to get the vehicle to do a max rate climb. That is the only purpose of the target altitude. The key thing is that this target is not used for the comparison of when the climb is done.

The problem is that without terrain, or if your at sea level then the target altitude may be lower than the threshold alt and the climb will never complete. 

We could just set `INT32_MAX` here and the result would be the same. Although I would not recommend that.

Reply:
without terrain it climbs to current amsl + climb alt in cm....so that's correct, right?
how do you suggest I set the max climb , terrain based?